overlord/devicestate/devicestate.go

```diff
@@ -967,45 +942,48 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 			// revision was unchanged. in this case, we treat the snap as if it
 			// were unchanged.
 			if ts.MaybeEdge(snapstate.LastBeforeLocalModificationsEdge) != nil {
-				if err := updateNeededSnapsFromTs(ts); err != nil {
-					return nil, err
-				}
-
 				continue
 			}
 		}
 
 		// if we're here, the snap that is installed is unchanged from the snap
 		// that the model requires. the snap may have had a channel change, but
 		// that channel change did not result in a revision change.
+		tracker.Add(currentInfo)
+	}
 
-		if currentInfo.Base != "" {
-			neededSnaps[currentInfo.Base] = true
+	for _, sn := range tracker.Snaps() {
+		gadgetBase := sn.Base
+		if gadgetBase == "" {
+			gadgetBase = "core"
 		}
-		// deal with content providers
-		for defProvider := range snap.NeededDefaultProviders(currentInfo) {
-			neededSnaps[defProvider] = true
+
+		modelBase := new.Base()
+		if modelBase == "" {
+			modelBase = "core"
 		}
-	}
-	// Now we know what snaps are in the model and whether they have any
-	// dependencies. Verify that the model is self contained, in the sense
-	// that all prerequisites of the snaps in the model, i.e. bases and
-	// default content providers are explicitly listed in the model
-	var missingSnaps []string
-	for needed := range neededSnaps {
-		if !snapsAccountedFor[needed] {
-			missingSnaps = append(missingSnaps, needed)
+
+		if sn.Type() == snap.TypeGadget && gadgetBase != modelBase {
```

Comment:
I'd instead do a "continue" at the beginning of the loop if not a gadget, then retrieve the gadget base

Reply:
Thanks, addressed this and pulled this bit out into a helper.